### PR TITLE
fix(graph): don't reject read-only Cypher that names properties like write keywords

### LINKED ISSUE #1375: db_neo4j: read-only keyword scan false-positives on keyword-named properties/labels
### Summary

`db_neo4j`'s read-only keyword scan (`_is_cypher_safe` in `nodes/src/nodes/db_neo4j/utils.py`) uses `\bKEYWORD\b`, so it **false-positives on read-only Cypher** that references a property, label, or relationship type named exactly like a write keyword (`CREATE`/`MERGE`/`DELETE`/`SET`/`REMOVE`/`DROP`/`FOREACH`/…). The valid read is then refused.

Example (read-only, wrongly rejected):

```cypher
MATCH (n) RETURN n.delete
```

This is the same class of bug fixed for `db_arango` in #1294 (maintainer review by @asclearuc raised it there and suggested tracking neo4j separately).

### Why neo4j needs a different fix than arango

`db_arango` could simply **lean on its EXPLAIN-plan gate**: `_validate_query` runs `EXPLAIN` *and* inspects the plan for data-modification node types, so it is a precise read-only authority and the keyword scan was demoted to a soft backstop.

`db_neo4j` does **not** have that: its `_validate_query` runs `EXPLAIN` for **syntax only** (`session.run(f'EXPLAIN {cypher}').consume()`) — there is no plan/operator inspection. So in neo4j the keyword scan **is** the primary read-only gate and cannot just defer to EXPLAIN.

### Suggested direction

1. **Minimum:** add the same negative look-behind to `_UNSAFE_CYPHER` so a keyword used as property access / compound identifier is not matched:
   ```python
   # e.g. (?<![\w.])(?:CREATE|MERGE|DELETE|DETACH\s+DELETE|SET|REMOVE|DROP|FOREACH|LOAD\s+CSV|...)\b
   ```
   This fixes the attribute-access case (`n.delete`) but not a bare label/identifier spelled like a keyword.
2. **Consider:** add a real read-only gate so label/identifier names spelled like keywords also pass — e.g. inspect Neo4j's EXPLAIN plan for write operators (`Create`, `Merge`, `Delete`, `SetProperty`, …), or run inside a read transaction / use `dbms.security`/`ACCESS` read-only routing. This mirrors what `db_arango` does with `_plan_is_modification`.

### Scope

`db_neo4j` only. Reference implementation for the arango side: #1294.

### PR DESCRIPTION
**Pull request by Ansh-Karnwal**
Fixes #1375

## Problem

The shared read-only gate `is_cypher_safe` (in `packages/ai/src/ai/common/graph/cypher_safety.py`) matches write keywords with `\bKEYWORD\b`. That word boundary also matches a keyword used as a property or map key, so a read-only query like:

```cypher
MATCH (n) RETURN n.delete
```

is wrongly rejected as a write. The gate also strips comments but never string literals, so a keyword that is only data (`WHERE n.op = "CREATE"`) trips it too. This is the same class of bug that was fixed for `db_arango` in #1294.

## Where this landed

I first fixed this in `db_neo4j`'s own `utils.py`. #1611 then migrated every graph node onto a shared base and removed that file, moving the gate into `cypher_safety.py`. This PR is re-pointed there, so the one fix now covers every graph node that calls `is_cypher_safe` (FalkorDB today, and any future driver on `graph_instance_base`, which calls it from three sites). Thanks to @dsapandora for flagging the migration.

## Change

- `_UNSAFE_CYPHER`: add a leading negative look-behind `(?<![\w.])` so a write keyword used as dotted access or a compound identifier (`n.delete`, `n.create`) is no longer read as the clause itself.
- `is_cypher_safe`: strip comments together with string and back-quoted-identifier literals in a single left-to-right pass. A keyword that appears only inside a string no longer causes a false rejection, and because the scan is single-pass, an unbalanced quote inside a comment can no longer swallow a real trailing write keyword (a bypass the previous chained strippers allowed).

A bare label or identifier spelled exactly like a keyword can still match, since a text scan cannot tell a keyword from an identifier. Cypher requires such identifiers to be back-quoted, and back-quoted spans are stripped before the scan, so the remaining gap is narrow. Closing it fully needs a plan-based read-only gate (FalkorDB already has one server-side via `GRAPH.RO_QUERY`); that is out of scope here.

## Tests

Adds `packages/ai/tests/ai/common/graph/test_cypher_safety.py` covering keyword-named properties, keywords that appear only as string or comment data, every write clause, and the comment/string bypass cases (including the unbalanced-quote-in-comment regression). The existing `test_graph_falkordb.py` assertions for this gate still pass.

```bash
pytest packages/ai/tests/ai/common/graph/test_cypher_safety.py
```

Read-only queries such as `MATCH (n) RETURN n.delete` now pass the gate; writes such as `MATCH (n) DETACH DELETE n` are still rejected.

## Summary by CodeRabbit

* **Bug Fixes**
  * Strengthened Cypher safety validation to better ignore write keywords found inside comments and string/identifier literals.
  * Improved robustness against malformed input that could previously hide unsafe clauses, including tricky line-comment edge cases.
  * Continued to allow valid read-only queries and data that happens to match write-related keywords.

* **Tests**
  * Added extensive automated coverage for safe reads, unsafe write/admin clauses, and bypass-resistant comment/literal scenarios.

### CODE REVIEW

File: nodes/src/nodes/db_neo4j/utils.py [hunk 1]
```diff
@@ -61,12 +68,20 @@ def _is_cypher_safe(cypher: str) -> bool:
...
     """
-    # Strip both single-line and block comments before checking.
-    stripped = re.sub(r'//[^\n]*', '', cypher)
+    # Strip string literals FIRST (handling \" / \' escapes), so a keyword or a
+    # // sequence inside a string can neither trigger a false rejection nor hide
+    # a real trailing write keyword. Then strip single-line and block comments.
+    stripped = re.sub(r'"(?:[^"\\]|\\.)*"', '', cypher)
+    stripped = re.sub(r"'(?:[^'\\]|\\.)*'", '', stripped)
+    stripped = re.sub(r'//[^\n]*', '', stripped)
     stripped = re.sub(r'/\*.*?\*/', '', stripped, flags=re.DOTALL)
```
[the remark this answers is not in the document]
Comment (Ansh-Karnwal) [reply nodes/src/nodes/db_neo4j/utils.py]: Good catch, and I confirmed it's exploitable locally. With the sequential stripping, the lone quote in the first comment made the double-quoted-string regex match across the newlines all the way to the quote in the trailing comment, so

```
MATCH (n)
// start "
DELETE n
// end "
```

collapsed to just `MATCH (n)` and passed the gate as read-only. I folded the four patterns into one alternation pass as suggested, so a comment consumes its own contents before the string rule can match a quote inside it. Pushed in 4c1945e0.

File: nodes/test/test_db_neo4j.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        assert _is_cypher_safe('MATCH (n) RETURN n // then DELETE later') is True
+        assert _is_cypher_safe('/* REMOVE everything */ MATCH (n) RETURN n') is True
+
+    def test_keyword_inside_string_literal_is_safe(self):
+        assert _is_cypher_safe("MATCH (n) WHERE n.status = 'DELETE' RETURN n") is True
+        assert _is_cypher_safe('MATCH (n) WHERE n.op = "CREATE" RETURN n') is True
+
+    def test_slashslash_in_string_does_not_hide_write_keyword(self):
+        query = 'MATCH (n) WHERE n.url = "http://x" DETACH DELETE n'
+        assert _is_cypher_safe(query) is False
```
[the remark this answers is not in the document]
Comment (Ansh-Karnwal) [reply nodes/test/test_db_neo4j.py]: Added, thanks. I used the unbalanced-quote-in-comment query as the regression case so this exact bypass stays covered.

**Comment by Ansh-Karnwal** [comment]
Addressed the CodeRabbit review in 4c1945e0: folded the string/comment stripping into a single alternation pass to close the safety-gate bypass, and added a regression test for the unbalanced-quote-in-comment case. Confirmed the gate now rejects the exploit query while still allowing keyword-named attributes/strings.

**Comment by dsapandora** [comment]
The fix itself is right and still needed, but `develop` moved under you: #1611 (merged after you opened this) migrated Neo4j onto the shared graph base, so `nodes/src/nodes/db_neo4j/utils.py` no longer exists — that's why the PR now shows `CONFLICTING`.

The exact same read-only gate now lives in **`packages/ai/src/ai/common/graph/cypher_safety.py`** (`is_cypher_safe` / `_UNSAFE_CYPHER`), and it still has both bugs you caught: the `\b(?:CREATE|DELETE|…)` word-boundary (matches `n.delete`), and comment-only stripping that doesn't strip string literals (`n.op = "CREATE"`). Could you re-point this PR there? Applying your `(?<![\w.])` look-behind + single-pass string/comment strip in that one file fixes it for **every** graph node at once — the gate is also called from `graph_instance_base.py` (lines 225, 341, 519), not just Neo4j.

For the test: `utils.py` was stdlib-only so you loaded it via `importlib`; the shared module is a normal import now, so `from ai.common.graph.cypher_safety import is_cypher_safe` works directly. There's already sibling coverage to match in `packages/ai/tests/ai/common/database/test_sql_safety.py` and `nodes/test/test_graph_falkordb.py`. Note your `_parse_is_valid` / `_strip_ns` cases don't port — those helpers moved onto the base during the migration.

Thanks for chasing this down — the false-positive analysis is spot on, it just needs to land one layer up.

**Comment by Ansh-Karnwal** [comment]
Done, re-pointed to `packages/ai/src/ai/common/graph/cypher_safety.py` and rebased onto current `develop`, so the old `db_neo4j/utils.py` changes drop out and the PR is mergeable again.

The fix carries over as you described: `(?<![\w.])` in place of the leading `\b` so `n.delete` / `n.create` read as property access, and a single-pass strip that now covers string and back-quoted-identifier literals alongside comments. So `n.op = "CREATE"` no longer trips the gate, and an unbalanced quote in a comment can't hide a trailing write keyword. Since every caller goes through this one function, the fix now covers the three `graph_instance_base` sites and any future driver, all from the one place.

Tests moved to `packages/ai/tests/ai/common/graph/test_cypher_safety.py` as a normal import, matching the `test_sql_safety.py` style. They cover keyword-named properties, keywords that appear only as string/comment data, every write clause, and the bypass cases including the unbalanced-quote-in-comment regression. I also re-ran the existing `test_graph_falkordb.py` gate assertions against the change and they still pass. Dropped the `_parse_is_valid` / `_strip_ns` cases since those helpers moved onto the base in the migration.

One residual, called out in the PR body: a bare identifier spelled exactly like a keyword still matches, but Cypher requires those to be back-quoted and back-quoted spans are stripped first, so the gap is narrow. The server-side `GRAPH.RO_QUERY` stays the real authority.

Thanks for the pointer.

File: packages/ai/src/ai/common/graph/cypher_safety.py [hunk 3]
```diff
@@ -39,12 +39,36 @@
...
+_STRIP_LITERALS = re.compile(
+    r"""
+      //[^\n]*             # line comment
+    | /\*.*?\*/            # block comment
+    | '(?:\\.|[^'\\])*'    # single-quoted string
+    | "(?:\\.|[^"\\])*"    # double-quoted string
+    | `[^`]*`              # back-quoted identifier
+    """,
+    re.DOTALL | re.VERBOSE,
+)
```
[the remark this answers is not in the document]
Comment (Ansh-Karnwal) [reply packages/ai/src/ai/common/graph/cypher_safety.py:74]: Good catch, fixed in d5cabcd2. Confirmed the bypass locally: `MATCH (n) // x\rDELETE n` passed the gate before and is rejected now, while a normal LF-terminated comment and read-only queries are unaffected. Changed the line-comment span to `//[^\r\n]*`.

Related heads-up, not for this PR: the SQL sibling has the same class of bug. `packages/ai/src/ai/common/database/sql_safety.py` strips line comments with `re.sub(r'--.*?$', '', sql, flags=re.MULTILINE)`, and `.` still consumes a lone `\r`, so `-- x\rDROP TABLE t` gets swallowed the same way. I left it out of scope here since it is a different module and ticket, but it likely wants the same fix.

File: packages/ai/tests/ai/common/graph/test_cypher_safety.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+@pytest.mark.parametrize(
+    'cypher',
+    [
+        'MATCH (n) /* harmless */ DELETE n',
+        '/* SELECT-ish */ MATCH (n) SET n.x = 1',
+        "MATCH (n {name: 'safe'}) DELETE n",
+        'MATCH (n) WHERE n.x = "CREATE" CREATE (m)',
+    ],
+)
```
[the remark this answers is not in the document]
Comment (Ansh-Karnwal) [reply packages/ai/tests/ai/common/graph/test_cypher_safety.py:131]: Added in d5cabcd2: `test_carriage_return_ends_line_comment` covers `// hide me\rDELETE n` and `//\rCREATE (m)`, both asserted rejected.

**Comment by nihalnihalani** [comment]
Both bugs are still present in the file after the re-point: [`cypher_safety.py:42`](https://github.com/rocketride-org/rocketride-server/blob/develop/packages/ai/src/ai/common/graph/cypher_safety.py#L42) still anchors on `\b(?:CREATE|MERGE|DELETE|...)`, so `MATCH (n) RETURN n.delete` is rejected, and `is_cypher_safe` strips only `//` and `/* */`, never string literals, so `WHERE n.op = "CREATE"` trips it too. The PR is mergeable and I found nothing blocking in the change.

Two notes for whoever picks it up:

- The build and Ruff jobs have never run on this head — the only checks present are Auto-label, CodeRabbit and discord-notify.
- The module docstring on develop already frames this as defence-in-depth behind a server-side guarantee (FalkorDB's `GRAPH.RO_QUERY`). #2118 is making the same argument for the SQL side, so it may be worth reading the two together.

**Review by stepmikhaylov — APPROVED** [review]
